### PR TITLE
[Integration][Custom] Add support for skip token pagination

### PR DESCRIPTION
**Pull request by dev-habib-nuhu**
### **User description**
# Description

**What** - Added skip token pagination support to the Custom HTTP integration, enabling seamless integration with APIs that use skip token-based pagination (e.g., Microsoft Graph API, Azure DevOps API, OData services).

**Why** - Many modern APIs, particularly Microsoft's Graph API and OData-compliant services, use skip token-based pagination instead of traditional offset/page-based approaches. Without this pagination type, users couldn't integrate these APIs using the Custom integration and had to resort to workarounds or custom solutions.

**How** - 
- Implemented a new `SkipTokenPagination` handler class that intelligently extracts skip tokens from API responses
- Added flexible configuration options to support various skip token implementations:
  - `skipTokenParam`: Query parameter name for skip token (default: "skiptoken")
  - `nextLinkPath`: Dot-notation path to extract next link URL from response (e.g., `"links.next"`, `"@odata.nextLink"`)
  - `skipTokenPath`: Dot-notation path to extract skip token directly from response (e.g., `"pagination.skipToken"`)
  - `sizeParam`: Query parameter for page size (e.g., `"limit"`, `"$top"`)
- Implemented automatic fallback to common field names (`nextLink`, `next_link`, `@odata.nextLink`, `nextToken`, `skipToken`, etc.)
- Added token extraction logic that parses skip tokens from URL query parameters
- Added comprehensive unit tests (11 test cases) covering various URL formats and edge cases
- Updated spec.yaml, README documentation with configuration examples and usage patterns

## Type of change

- [x] New feature (non-breaking change which adds functionality)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Add skip token pagination handler supporting Microsoft Graph and OData APIs

- Implement flexible token extraction from URLs and response fields

- Add 11 comprehensive unit tests for URL parsing and edge cases

- Update spec.yaml and README with skip token configuration examples

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Custom HTTP Integration"] -->|"Pagination Type"| B["SkipTokenPagination Handler"]
  B -->|"Extract Token"| C["From Next Link URL"]
  B -->|"Extract Token"| D["From Response Field"]
  C -->|"Parse Query Params"| E["Skip Token"]
  D -->|"Nested Path Lookup"| E
  E -->|"Next Request"| F["API with Skip Token"]
  F -->|"Response"| B
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>handlers.py</strong><dd><code>Implement skip token pagination handler with flexible extraction</code></dd></summary>
<hr>

integrations/custom/http_server/handlers.py

<ul><li>Implement new <code>SkipTokenPagination</code> class extending <code>PaginationHandler</code><br> <li> Add <code>_extract_token_from_url()</code> method supporting multiple token <br>parameter formats ($skiptoken, skiptoken, skip_token, skipToken)<br> <li> Support configurable token extraction from response fields via <br>dot-notation paths<br> <li> Implement fallback logic for common field names (nextLink, next_link, <br>@odata.nextLink, nextToken, skipToken)<br> <li> Register skip_token handler in PAGINATION_HANDLERS registry</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-02dc3c095592539a02313c909d4101d589faf4ba1be99db18f8bae910329a40b">+83/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_pagination_handlers.py</strong><dd><code>Add comprehensive unit tests for skip token pagination</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/tests/test_pagination_handlers.py

<ul><li>Add 11 test cases for <code>SkipTokenPagination._extract_token_from_url()</code> <br>method<br> <li> Test OData format ($skiptoken), standard format (skiptoken), <br>snake_case (skip_token), and camelCase (skipToken) token extraction<br> <li> Test URL parsing with multiple query parameters and edge cases <br>(invalid URL, empty string, no token)<br> <li> Test configuration defaults and custom parameter names</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-b2b7eee0402f82eb3c3c305e9ec61167208b8a2e017c2a056382658fba78607a">+96/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>spec.yaml</strong><dd><code>Update spec with skip token pagination configuration options</code></dd></summary>
<hr>

integrations/custom/.port/spec.yaml

<ul><li>Add "skip_token" to paginationType enum and update description<br> <li> Update pageSize, paginationParam, sizeParam, and hasMorePath <br>dependencies to include skip_token type<br> <li> Add three new configuration fields: skipTokenParam, nextLinkPath, <br>skipTokenPath with conditional dependencies on skip_token pagination <br>type<br> <li> Update sizeParam description to include $top example for OData APIs<br> <li> Add customHeaders configuration field for custom HTTP headers support</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-c428ac234d25ce406220e723ad1ff477fe4ccfedeae9670a3662a8af37b3779d">+43/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Bump version to 0.2.42-beta</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/pyproject.toml

- Update version from 0.2.41-beta to 0.2.42-beta

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-3ccbf74a6831bd381a6d5e8a40d5db2648ebdfb25a370e7aab4a5075440fcb18">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update changelog with new feature version</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/CHANGELOG.md

- Add version 0.2.42-beta entry with skip token pagination feature

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-fb2baecf0e2f18c4437d00a27df028ce92bcc3978262d29691fc1a2604b4f96a">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add skip token pagination documentation and examples</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/README.md

<ul><li>Update feature list to include skip token pagination option<br> <li> Add skip token pagination environment variable examples<br> <li> Add comprehensive skip token pagination section with three usage <br>examples (generic, direct field, Microsoft Graph)<br> <li> Document configuration options for next_link_path, skip_token_path, <br>skip_token_param, and size_param<br> <li> Explain fallback behavior for common field names<br> <li> Fix minor formatting issues (trailing spaces, line breaks)</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2629/files#diff-9163c9f3e6e3b161969886be5f610784711ba58fa2cbc6479dd0a4b60469c581">+81/-15</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/custom/http_server/handlers.py [hunk 1]
```diff
@@ -303,12 +305,93 @@ async def fetch_all(
...
+                        or response_data.get("@odata.nextLink")
+                    )
+                    if next_link:
+                        next_skip_token = self._extract_token_from_url(next_link)
+                    if not next_skip_token:
+                        next_skip_token = (
+                            response_data.get("skipToken")
+                            or response_data.get("skip_token")
+                            or response_data.get("nextToken")
+                        )
```
Comment (oiadebayo) [inline integrations/custom/http_server/handlers.py]: Rather than using .get() with this limited option, can we expose a variable for users to a supply the path of the next link and the next skip token
